llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -158,10 +173,48 @@ void insertMinIterAssumption(ICmpInst &LoopCmpInstr, BasicBlock &LoopHeader,
-                                               : LoopCmpInstr.getOperand(0);
-  Value *RHS = BunldedSCEV.IVOperandIndex == 0 ? LoopCmpInstr.getOperand(1)
-                                               : LoopAssertValue;
+  Value *Bound = BunldedSCEV.IVOperandIndex == 0 ? LoopCmpInstr.getOperand(1)
+                                                 : LoopCmpInstr.getOperand(0);
+
+  // Decouple the Bound and the ICmp Instruction by inserting 2 reversable
+  // Adds. If the Bound consists of a Load Instruction followed by a Constant
+  // offset, Correlated Value Propagation can deduce
+  // from the Assumption what the Branch Condition will evaluate to.           \
+  //       Conditional branch non-loop case \
+  // %i0 = add i32 0, 0                                                        \
+  // %0 = load i32, ptr p                                                      \
+  // %sub0 = add i32, %0, -1                                                   \
+  // %cmp0 = icmp ult i32 %sub0, 4 \
+  // call void @llvm.assume(i1 %cmp0)                                          \
+  // %cmp = icmp ult %i0, %sub0                                                \
+  // br i1 %cmp, label %if.true, label %if.false                               \
+  // In the Conditional Branch Case (non Loop), the Branch Condition is also
+  // set to the appropriate Boolean, so that the assumption is safe to remove
+  //  and it is removed, see code below \
+  // %i0 = add i32 0, 0 \
+  // %0 = load i32, ptr p                                                      \
+  // %sub0 = add i32, %0, -1                                                   \
+  // br true, label %if.true, label %if.false                                  \
+  // With a Loop Branch Condition, the Boolean in the Branch Instruction cannot
+  // be inserted, because an exit condition is needed. However, the assumption
+  // is still removed. To mitigate this, the link between the Assumption and
+  // the Load Instruction has to be so complex that Correlated Value
+  // Propagation cannot deduce any information from the Assumption about the
+  // Load Value range. The following structure is sufficient and is similar to
+  // what builtin_assume inserts.
+  // %i0 = add i32 0, 0                                                        \
+  // %0 = load i32, ptr p                                                      \
+  // %sub0 = add i32, %0, -1                                                   \
+  // %decouple0 = add i32, %0, 1                                               \
+  // %decouple1 = add i32, %decouple0, -1                                      \
+  // %cmp0 = icmp ult i32 %sub0, 4                                             \
+  // call void @llvm.assume(i1 %cmp0)                                          \
+  //  Overflow/Underflow considerations are not needed, because the Instruction
+  //  Combine Pass will handle them.
+  Bound = insertBufferAdds(Bound, Builder);
```

Comment:
This is a very long comment, would a fix of the CorrelatedValue pass fit in less lines? :)

Reply:
The weird thing is, that `builtin_assume `is doing exactly the same thing.... (by changing instructions and going through an Instruction Combine pass). 

However,  we cannot mirror this approach, since it is even more fragile and is not consistent enough. I assume our position in the optimization pipeline is not "correct" for it to work.